code/sha3/Hacl.Impl.SHA3.Vec.fst

```diff
@@ -0,0 +1,1503 @@
+  let b0 = b.(|0|) in
+  next_blocks rateInBytes b0;
+  let h1 = ST.get() in
+  Lib.NTuple.eq_intro (as_seq_multi h1 b) 
+    (V.next_block1 #m (v rateInBytes) (as_seq_multi h0 b))
+
+inline_for_extraction noextract
+val next_block4: #m:m_spec{lanes m == 4}
+  -> rateInBytes:size_t{v rateInBytes > 0 /\ v rateInBytes <= 256}
+  -> b:multibuf (lanes m) 256ul ->
+  Stack unit
+  (requires fun h -> live_multi h b /\ internally_disjoint b /\ 
+    as_seq_multi h b == next_block_seq_zero m)
+  (ensures  fun h0 _ h1 -> modifies_multi b h0 h1 /\
+    as_seq_multi h1 b == V.next_block #m (v rateInBytes) (as_seq_multi h0 b))
+
+let next_block4 #m rateInBytes b =
+  let h0 = ST.get() in
+  let (b0,(b1,(b2,b3))) = tup4 b in
+  assert (internally_disjoint4 b0 b1 b2 b3);
+  next_blocks rateInBytes b0;
+  next_blocks rateInBytes b1;
+  next_blocks rateInBytes b2;
+  next_blocks rateInBytes b3;
+  loc_multi4 b;
+  let h1 = ST.get() in
+  Lib.NTuple.eq_intro (as_seq_multi h1 b) 
+    (V.next_block4 #m (v rateInBytes) (as_seq_multi h0 b))
+
+inline_for_extraction noextract
+val next_block: #m:m_spec{is_supported m}
+  -> rateInBytes:size_t{v rateInBytes > 0 /\ v rateInBytes <= 256}
+  -> b:multibuf (lanes m) 256ul ->
+  Stack unit
+  (requires fun h -> live_multi h b /\ internally_disjoint b /\ 
+    as_seq_multi h b == next_block_seq_zero m)
+  (ensures  fun h0 _ h1 -> modifies_multi b h0 h1 /\
+    as_seq_multi h1 b == V.next_block #m (v rateInBytes) (as_seq_multi h0 b))
+
+let next_block #m rateInBytes b =
+  match lanes m with
+  | 1 -> next_block1 #m rateInBytes b
+  | 4 -> next_block4 #m rateInBytes b
+
+inline_for_extraction noextract
+val alloc_multi1: m:m_spec{lanes m == 1} ->
+  StackInline (multibuf (lanes m) 256ul)
+  (requires fun h -> True)
+  (ensures  fun h0 b h1 -> live_multi h1 b /\
+    stack_allocated_multi b h0 h1 (Seq.create 256 (u8 0)) /\
+    as_seq_multi h1 b == next_block_seq_zero m)
+
+let alloc_multi1 m =
+  let b = create 256ul (u8 0) in
+  let b = ntup1 b in
+  let h0 = ST.get() in
+  Lib.NTuple.eq_intro (as_seq_multi h0 b) (next_block_seq_zero m);
+  b
+
+inline_for_extraction noextract
+val alloc_multi4: m:m_spec{lanes m == 4} ->
+  StackInline (multibuf (lanes m) 256ul)
+  (requires fun h -> True)
+  (ensures  fun h0 b h1 -> live_multi h1 b /\ internally_disjoint b /\
+    stack_allocated_multi b h0 h1 (Seq.create 256 (u8 0)) /\
+    as_seq_multi h1 b == next_block_seq_zero m)
+
+let alloc_multi4 m =
+  //let h0' = ST.get() in
+  let b0 = create 256ul (u8 0) in
+  let b1 = create 256ul (u8 0) in
+  let b2 = create 256ul (u8 0) in
+  let b3 = create 256ul (u8 0) in
+  let b = ntup4 (b0, (b1, (b2, b3))) in
+  let h0 = ST.get() in
+  Lib.NTuple.eq_intro (as_seq_multi h0 b) (next_block_seq_zero m);
+  //let h1 = ST.get() in
+  //assert (stack_allocated b0 h0' h1 (Seq.create 256 (u8 0)));
+  //assert (stack_allocated b1 h0' h1 (Seq.create 256 (u8 0)));
+  admit(); //Can't prove (stack_allocated_multi b h0 h1 (Seq.create 256 (u8 0)))
```

Comment:
is this still an issue?

Reply:
I forget about this one, I will update the relative functions shortly.

Reply:
The issue is fixed, admits removed with the last commit.